control-plane/agents/core/src/core/scheduling/volume.rs

```diff
@@ -241,137 +251,108 @@ impl GetChildForRemovalContext {
-        let replicas = self.registry.specs.get_volume_replicas(&self.vol_spec.uuid);
-        let mut replicas = replicas.iter().map(|r| r.lock().clone());
-
-        self.nexus_spec
-            .children
-            .iter()
-            .map(|child| {
-                let spec = child
-                    .as_replica()
-                    .map(|r| replicas.find(|s| &s.uuid == r.uuid()))
-                    .flatten();
-                let child_state = nexus
-                    .as_ref()
-                    .map(|n| n.children.iter().find(|c| c.uri == child.uri()));
-
-                NexusChildItem::new(spec, child.uri(), child_state.flatten())
-            })
-            .collect::<Vec<_>>()
+impl ReplicaRemovalCandidates {
+    /// Get the next healthy replica removal candidate
+    fn next_healthy(&mut self) -> Option<ReplicaItem> {
+        let replica_count = self.context.spec.target_num_replicas();
+        let healthy_online = self.healthy.iter().filter(|replica| match replica.state() {
+            None => false,
+            Some(state) => state.online(),
+        });
+        // removing too many healthy_online replicas could compromise the volume's redundancy
+        if healthy_online.into_iter().count() > replica_count as usize {
+            match self.healthy.pop() {
+                Some(replica) if self.context.unused_only & replica.child_spec().is_none() => {
+                    Some(replica)
+                }
+                replica => replica,
+            }
+        } else {
+            None
+        }
+    }
+    /// Get the next unhealthy candidates (any is a good fit)
+    fn next_unhealthy(&mut self) -> Option<ReplicaItem> {
+        self.unhealthy.pop()
+    }
+    /// Get the next removal candidate.
+    /// Unhealthy replicas are removed before healthy replicas
+    pub fn next(&mut self) -> Option<ReplicaItem> {
+        self.next_unhealthy().or_else(|| self.next_healthy())
     }
-}
 
-impl DecreaseNexusReplica {
-    async fn builder(vol_spec: &VolumeSpec, nexus_spec: &NexusSpec, registry: &Registry) -> Self {
-        let context = GetNexusChildForRemovalContext {
-            registry: registry.clone(),
-            vol_spec: vol_spec.clone(),
-            nexus_spec: nexus_spec.clone(),
+    fn new(context: GetChildForRemovalContext, items: Vec<ReplicaItem>) -> Self {
+        let has_info = context.nexus_info.is_some();
+        let is_healthy = |item: &&ReplicaItem| -> bool {
+            match item.child_info() {
+                Some(info) => info.healthy,
+                None => !has_info,
+            }
         };
+        let is_not_healthy = |item: &&ReplicaItem| -> bool { !is_healthy(item) };
         Self {
-            list: context.list().await,
             context,
+            // reverse so we can pop them
+            healthy: items.iter().filter(is_healthy).rev().cloned().collect(),
+            unhealthy: items.iter().filter(is_not_healthy).rev().cloned().collect(),
```

Comment:
Because we don't iterate through the replicas with this type, we pop them, and pop removes the last element, so now the "less preferred" replicas need to be at the back. I'll add a comment as well.

Reply:
oh I misread your comment. No not all replicas are equal, they are sorted according to a set of rules. Ex, node locality, state, etc..